libr/anal/type.c

```diff
@@ -122,18 +122,108 @@ R_API void r_anal_save_parsed_type(RAnal *anal, const char *parsed) {
 static int typecmp(const void *a, const void *b) {
 	char *type1 = (char *) a;
 	char *type2 = (char *) b;
-	return strcmp (type1 , type2);
+	return strcmp (type1, type2);
 }
 
 R_API RList *r_anal_types_from_fcn(RAnal *anal, RAnalFunction *fcn) {
 	RListIter *iter;
 	RAnalVar *var;
 	RList *list = r_anal_var_all_list (anal, fcn);
-        RList *type_used = r_list_new ();
-        r_list_foreach (list, iter, var) {
-		r_list_append (type_used , var->type);
-        }
-	RList *uniq = r_list_uniq (type_used , typecmp);
+	RList *type_used = r_list_new ();
+	r_list_foreach (list, iter, var) {
+		r_list_append (type_used, var->type);
+	}
+	RList *uniq = r_list_uniq (type_used, typecmp);
 	r_list_free (type_used);
 	return uniq;
 }
+
+static void enum_type_fini(void *e, void *user) {
+	(void)user;
+	RAnalEnumCase *cas = e;
+	free ((char *)cas->name);
+}
+
+static RAnalBaseType *get_enum_type(RAnal *anal, const char *sanitized_name) {
+	assert (sanitized_name != NULL);
+
+	RAnalBaseType *base_type = malloc (sizeof(RAnalBaseType));
+	if (!base_type) {
+		return NULL;
+	}
+	base_type->kind = R_ANAL_BASE_TYPE_KIND_ENUM;
+	
+	RAnalBaseTypeEnum base_enum;
+	
+	char *key = sdb_fmt ("%s.%s", "enum", sanitized_name);
+	char *members = sdb_get (anal->sdb_types, key, NULL);
+	RVector *cases = r_vector_new (sizeof(RAnalEnumCase), enum_type_fini, NULL);
+	if (!cases) {
+		free (base_type);
+		return NULL;
+	}
+	base_enum.cases = cases;
+
+	if (!r_vector_reserve (cases, (size_t) sdb_alen (members))) {
+		free (base_type);
+		r_vector_free (cases);
+		return NULL;
+	}
+
+	char *cur;
+	void *element;
+	sdb_aforeach (cur, members) {
+		const char *value = sdb_get (anal->sdb_types, sdb_fmt ("%s.%s.%s", "enum", sanitized_name, cur), NULL);
+		RAnalEnumCase cas = {.name = strdup (cur), .val = strtol (value, NULL, 16)};
+		free (value);
+		
+		element = r_vector_push (cases, &cas); // returns null if no space available
```

Comment:
Just a small tip: you can pass `NULL` to the push, which will allocate the new element, but not initialize it. Then you can initialize it in-place and avoid a memcpy. Whether you want to do that or not depends on the case.